h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1748,102 +1765,120 @@ private int getMovePriority(Chunk chunk) {
-            for (Iterator<Chunk> iter = move.iterator(); iter.hasNext(); ) {
-                Chunk chunk = iter.next();
-                if (fileStore.predictAllocation(chunk.len) >= leftmostBlock) {
-                    break;
-                }
-                moveChunk(chunk, false);
-                iter.remove();
-            }
 
+            Iterator<Chunk> iterator = move.iterator();
+            assert iterator.hasNext();
+            long leftmostBlock = iterator.next().block;
+            long originalBlokCount = getAfterLastBlock();
+            // we need to ensure that chunks moved within the following loop
+            // do not overlap with space just released by chunks moved before them,
+            // hence the need to reserve this area [leftmostBlock, originalBlokCount)
             for (Chunk chunk : move) {
-                moveChunk(chunk, true);
+                moveChunk(chunk, leftmostBlock, originalBlokCount);
             }
-
-            // update the metadata (store at the end of the file)
-            reuseSpace = false;
-            commit();
+            // update the metadata (hopefully within the file)
+            store(leftmostBlock, originalBlokCount);
             sync();
 
-            Chunk chunk = lastChunk;
+            Chunk chunkToMove = lastChunk;
+            long postEvacuationBlockCount = getAfterLastBlock();
 
-            // now re-use the empty space
-            reuseSpace = true;
+            boolean movedToEOF = chunkToMove.block >= originalBlokCount;
+            // move all chunks, which previously did not fit before reserved area
+            // now we can re-use previously reserved area [leftmostBlock, originalBlokCount),
+            // but need to reserve [originalBlokCount, postEvacuationBlockCount)
             for (Chunk c : move) {
-                moveChunk(c, false);
+                if (c.block >= originalBlokCount) {
+                    moveChunk(c, originalBlokCount, postEvacuationBlockCount);
+                    assert c.block < originalBlokCount;
+                    movedToEOF = true;
+                }
             }
+            assert postEvacuationBlockCount >= getAfterLastBlock();
 
-            // update the metadata (within the file)
-            commit();
-            sync();
-            if (moveChunk(chunk, false)) {
-                commit();
+            if (movedToEOF) {
+                boolean moved = moveChunkInside(chunkToMove, originalBlokCount);
+
+                // store a new chunk with updated metadata (hopefully within a file)
+                store(originalBlokCount, postEvacuationBlockCount);
+                sync();
+                // and if last chunk did not fit within a file, since we can use
+                // previously reserved area [originalBlokCount, postEvacuationBlockCount) now,
+                // try to move it closer to BOF
+                moved = !moved && moveChunkInside(chunkToMove, postEvacuationBlockCount);
```

Comment:
@andreitokar This is the last problem. In line 1814, chunkToMove.len always is less than postEvacuationBlockCount, so here either we remove the moveChunkInside(), or give **postEvacuationBlockCount - chunkToMove.len** instead of **postEvacuationBlockCount**, and the following "if" block need to make corresponding adjustment.

Reply:
@little-pan, Yes, you are absolutely right, that chunkToMove, under certain condition may stay where it is, leaving some hole near the end of file. This is is probably not a big deal really, because there where no overlap (no chance for corruption with ordered writes). Actually, unless were are talking about maintance on database shutdown, it may be not desirable to shrink the file right away, because it'll probably grow again next second. But this is totally separate discussion topic.
Anyway, I fixed it the way you suggested, except instead of subtracting, just use chunkToMove.block as the beginning of reserved area.